### ISSUE TITLE
[Bug][ROCm][Distributed] Custom all-reduce start barrier is fully relaxed — peer-data reads unordered against remote (XGMI) writes → torn reads

### ISSUE THREAD

**Issue by curvedinf** [vllm-project/vllm#57059]
### Your current environment

- Upstream vLLM @ `241e9391ed85f6c967432135427717a529776a63` (2026-09-15), ROCm 7.0 clang for codegen evidence
- 4× AMD Instinct MI100 (gfx908, XGMI full mesh), ROCm 7.14, torch 2.11.0+rocm7.1

### Current behavior

On multi-GPU ROCm where P2P goes over XGMI (on 4× MI100 / gfx908, ROCm 7.14; where the memory-model argument applies wherever peers are remote agents relative to device scope), the custom all-reduce kernels can pass their start barrier and read a peer's staged input buffer while the peer's writes are still in flight. Torn bf16 values seed NaN/huge-finite activations; in-engine this presents as episodic single-layer corruption and, at long context, junk-token output walls. Load- and timing-dependent; silent.

### Root cause (verified at HEAD `241e9391e`)

- `csrc/custom_collective_common.cuh` `barrier_at_start` (ROCm branch, lines 264–280): flag store to peers is `__ATOMIC_RELAXED` / `__MEMORY_SCOPE_SYSTEM`; flag load is `__ATOMIC_RELAXED` / `__MEMORY_SCOPE_DEVICE`. Fully relaxed — no release covers the producer's prior data stores, no acquire covers the consumer's subsequent data loads.
- `barrier_at_start_release` (lines 283–296) and non-final `barrier_at_end` (298–318) carry release/acquire, but **no kernel at HEAD calls `barrier_at_start_release`**: both `cross_device_reduce_1stage` (`csrc/custom_all_reduce.cuh:16`) and `cross_device_reduce_2stage` (line 51) synchronize the reduction start — the point that must order peer-data reads against peer-data writes — with the fully relaxed variant.
- SASS (gfx908, `-O3`, ROCm 7.0 clang): `barrier_at_start` emits a plain flag store and a bare spin (`global_load_dword ... glc` + `s_waitcnt vmcnt(0)`) with **zero fences**; the acquire variants emit `buffer_wbinvl1_vol`. Note: patching only the load scopes DEVICE→SYSTEM produced byte-identical device code on this toolchain — the **ordering** (relaxed → release/acquire) is the operative defect; scope promotion is required memory-model hygiene for remote-agent visibility but is codegen-inert on gfx908/ROCm 7.0.

### How to reproduce

Deterministic (no GPU, any machine with the repo + ROCm toolchain):

```bash
cd csrc && cat > /tmp/barriers.cu <<'EOF'
#include "custom_all_reduce.cuh"
__global__ void k_start(vllm::RankSignals sg, vllm::Signal* self, int rank) {
  vllm::barrier_at_start<4>(sg, self, rank);
}
__global__ void k_start_release(vllm::RankSignals sg, vllm::Signal* self, int rank) {
  vllm::barrier_at_start_release<4>(sg, self, rank);
}
EOF
clang++ -x hip --rocm-path=/opt/rocm --offload-arch=gfx908 -O3 -DUSE_ROCM \
  -I. -S /tmp/barriers.cu -o /tmp/barriers.s
# k_start spin loop: global_load glc + s_waitcnt, NO buffer_wbinvl1_vol
# k_start_release spin loop: same load + buffer_wbinvl1_vol (acquire)
grep -c buffer_wbinvl1_vol /tmp/barriers.s   # >0, but only from k_start_release
# And: neither reduce kernel at HEAD instantiates the acquire variant:
grep -n "barrier_at_start" custom_all_reduce.cuh   # both call the relaxed one
```

In-engine (4× MI100, XGMI): serve any TP=4 bf16 model with custom all-reduce on; drive long decode with speculative decoding at high concurrency, interleaving prefill chunks so all-reduces of many sizes run back-to-back; capture logits/hidden states across ranks vs a gather reference — episodic one-layer divergences (torn residuals); with enough steps, outputs collapse to token loops. Full protocol and knife-switch bisection: curvedinf/int8-vllm `docs/recipes/bug_rocm_batch_memcpy_graph_replay_race.md` and `PROBABLE_UPSTREAM_BUGS.md` BUG 1 (fix soak there: 24/24 long greedy legs after the fix vs ~5/8 boot failures before).

### How to characterize the bug as fixed

1. **Static/SASS (deterministic):** after the fix, both `cross_device_reduce_*` kernels' start-barrier spin loops emit `buffer_wbinvl1_vol` (acquire) before `__syncthreads()`, and all cross-rank flag loads/stores are `__MEMORY_SCOPE_SYSTEM`. Rebuild the two kernels and grep as above — the fences must now appear in the **reduce kernels** (not just the unused helper).
2. **Soak (rig):** the in-engine protocol above runs clean for ≥24/24 long greedy legs with no cross-rank one-layer divergence at matched positions.
3. **Unit:** a two-kernel spin-handshake test where the producer writes data then a flag with release, consumer acquires the flag then reads data, run under extreme SM contention, never observes stale data (a starting-point skeleton is attached: `kernels_barriers.cu`).

### Example solution

```diff
--- a/csrc/custom_all_reduce.cuh
+++ b/csrc/custom_all_reduce.cuh
@@ cross_device_reduce_1stage @@
-  barrier_at_start<ngpus>(sg, self_sg, rank);
+  barrier_at_start_release<ngpus>(sg, self_sg, rank);
@@ cross_device_reduce_2stage @@
-  barrier_at_start<ngpus>(sg, self_sg, rank);
+  barrier_at_start_release<ngpus>(sg, self_sg, rank);
--- a/csrc/custom_collective_common.cuh   (ROCm branch only)
+++ b/csrc/custom_collective_common.cuh
@@ barrier_at_start, barrier_at_start_release, barrier_at_end @@
-                                  __MEMORY_SCOPE_DEVICE) < flag);
+                                  __MEMORY_SCOPE_SYSTEM) < flag);
 (all three acquire/relaxed flag loads)
```

Verified: applying exactly this to the HEAD headers and rebuilding for gfx908, the 1-stage kernel's spin region goes from fences `NONE` to `['buffer_wbinvl1_vol']` (disassembly attached: `kernels_gfx908.s` = stock, `kernels_gfx908_release.s` = patched; the patched build's header `custom_all_reduce_release.cuh` is also attached). Measured cost on gfx908: none (the acquire invalidation is the only added instruction; XGMI paths are already coherent).

### Referenced code / evidence / example fix
[code.zip](https://github.com/user-attachments/files/32255309/code.zip)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
